benchmarks/osb/extensions/runners.py

```diff
@@ -0,0 +1,118 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+from .util import parse_int_parameter, parse_string_parameter
+import time
+
+
+def register(registry):
+    registry.register_runner(
+        "custom-vector-bulk", BulkVectorsFromDataSetRunner(), async_runner=True
+    )
+    registry.register_runner(
+        "custom-refresh", CustomRefreshRunner(), async_runner=True
+    )
+    registry.register_runner(
+        "train-model", TrainModelRunner(), async_runner=True
+    )
+    registry.register_runner(
+        "delete-model", DeleteModelRunner(), async_runner=True
+    )
+
+
+class BulkVectorsFromDataSetRunner:
+
+    async def __call__(self, opensearch, params):
+        size = parse_int_parameter("size", params)
+        retries = parse_int_parameter("retries", params, 0) + 1
+
+        for _ in range(retries):
+            try:
+                await opensearch.bulk(
+                    body=params["body"],
+                    timeout='5m'
+                )
+
+                return size, "docs"
+            except:
+                pass
+
+        raise TimeoutError("Failed to submit bulk request in specified number "
+                           "of retries: {}".format(retries))
+
+    def __repr__(self, *args, **kwargs):
+        return "custom-vector-bulk"
+
+
+class CustomRefreshRunner:
+
+    async def __call__(self, opensearch, params):
+        retries = parse_int_parameter("retries", params, 0) + 1
+
+        for _ in range(retries):
+            try:
+                await opensearch.indices.refresh(
+                    index=parse_string_parameter("index", params)
+                )
+
+                return
+            except:
+                pass
+
+        raise TimeoutError("Failed to refresh the index in specified number "
+                           "of retries: {}".format(retries))
+
+    def __repr__(self, *args, **kwargs):
+        return "custom-refresh"
+
+
+class TrainModelRunner:
+
+    async def __call__(self, opensearch, params):
+        # Train a model and wait for it training to complete
+        body = params["body"]
+        timeout = parse_int_parameter("timeout", params)
```

Comment:
it looks like retry than timeout. If we want timeout, then we should start timer and check value less than time out before iterating.

Reply:
Not sure I understand the proposal. I set it as timeout because the call returns before training finishes. So once the call is submitted, I check every second if the model state is created.

Reply:
Lets say timeout is 10 seconds. At line 73, 1 second is elapsed, but you are not accounting number of seconds it takes to execute from line 74 till 83. I believe it is more than a second that it is spent inside the loop. First, i will check is it possible to pass timeout to request itself and wait for the response ( or timeout exception from API itself). If that is not possible, we need a stop watch at line 71, and checks whether stop watch reached expected seconds as loop condition.

Reply:
I think the request itself isnt timing out. We are basically waiting on training to finish. I can setup the stop watch in the  loop.